File: frontend/src/util/sprites.ts
import {Game} from "../data/games/games.model";

export function getSpriteUrl(speciesName: string, game: Game, shiny: boolean) {
    const gameKeys = new Map<string, string>([
        ["RED", "red-blue"],
        ["BLUE", "red-blue"],
        ["YELLOW", "yellow"],
        ["SILVER", "silver"],
        ["GOLD", "gold"],
        ["CRYSTAL", "crystal"],
        ["RUBY", "ruby-sapphire"],
        ["SAPPHIRE", "ruby-sapphire"],
        ["FIRERED", "firered-leafgreen"],
        ["LEAFGREEN", "firered-leafgreen"],
        ["EMERALD", "emerald"],
        ["DIAMOND", "diamond-pearl"],
        ["PEARL", "diamond-pearl"],
        ["PLATINUM", "platinum"],
        ["HEARTGOLD", "heartgold-soulsilver"],
        ["SOULSILVER", "heartgold-soulsilver"],
        ["BLACK", "black-white"],
        ["WHITE", "black-white"],
        ["BLACK_2", "black-white"],
        ["WHITE_2", "black-white"],
        ["X", "x"],
        ["Y", "y"],
        ["OMEGA_RUBY", "omega-ruby-alpha-sapphire"],
        ["ALPHA_SAPPHIRE", "omega-ruby-alpha-sapphire"]
    ])
    const gameKey = gameKeys.get(game.key) ?? (game.generation === 7 ? "bank" : "home")
    const shinyKey = shiny ? "shiny" : "normal"
    const colorExtra = game.generation == 1 ? "-color" : ""
    return `https://img.pokemondb.net/sprites/${gameKey}/${shinyKey}/${speciesName.toLowerCase()}${colorExtra}.png`
}

File: backend/src/main/kotlin/net/namibsun/nuztrack/util/Sprites.kt
package net.namibsun.nuztrack.util

import net.namibsun.nuztrack.constants.enums.Games
import net.namibsun.nuztrack.transfer.PokemonSpeciesTO

fun getSpriteForGameAndPokemon(pokemon: PokemonSpeciesTO, game: Games, shiny: Boolean): String {
    val gameKeys = mapOf(
            Games.RED to "red-blue",
            Games.BLUE to "red-blue",
            Games.YELLOW to "yellow",
            Games.SILVER to "silver",
            Games.GOLD to "gold",
            Games.CRYSTAL to "crystal",
            Games.RUBY to "ruby-sapphire",
            Games.SAPPHIRE to "ruby-sapphire",
            Games.FIRERED to "firered-leafgreen",
            Games.LEAFGREEN to "firered-leafgreen",
            Games.EMERALD to "emerald",
            Games.DIAMOND to "diamond-pearl",
            Games.PEARL to "diamond-pearl",
            Games.PLATINUM to "platinum",
            Games.HEARTGOLD to "heartgold-soulsilver",
            Games.SOULSILVER to "heartgold-soulsilver",
            Games.BLACK to "black-white",
            Games.WHITE to "black-white",
            Games.BLACK_2 to "black-white",
            Games.WHITE_2 to "black-white",
            Games.X to "x-y",
            Games.Y to "y-y",
            Games.OMEGA_RUBY to "omega-ruby-alpha-sapphire/dex",
            Games.ALPHA_SAPPHIRE to "omega-ruby-alpha-sapphire/dex"
    )
    var gameKey = gameKeys[game]
    if (game in listOf(Games.FIRERED, Games.LEAFGREEN) && pokemon.pokedexNumber > 151) {
        gameKey = gameKeys[Games.EMERALD]
    }
    if (gameKey == null) {
        gameKey = if (game.generation == 7) "bank" else "home"
    }

    val shinyKey = if (shiny) "shiny" else "normal"
    val colorExtra = if (game.generation == 1) "-color" else ""
    return "https://img.pokemondb.net/sprites/" +
            "${gameKey}/${shinyKey}/" +
            pokemon.name.lowercase() +
            "${colorExtra}.png"
}
